webhooks/github: Support renaming topics on issue edit.

DhruvShetty22 opened the pull request:
This PR updates the GitHub webhook integration to automatically rename the corresponding Zulip topic when a Pull Request or Issue title is edited on GitHub. It uses the `check_topic_rename` helper to ensure that the entire conversation history is preserved and moved to the new topic, rather than creating a new topic.

- **Testing:**
    - Added unit tests for the renaming logic.
    - Added integration tests using the  `pull_request__edited`  and `issues__edited`fixture.
Fixes #34074

**How changes were tested:**

I verified this locally with both automated tests and manual integration testing.

**Automated:**
- Added a new backend test case to verify that `check_topic_rename` correctly identifies the old topic and updates the database.

**Manual (Test via ngrok):**
I connected a test GitHub repository to my local Zulip server using `ngrok` to verify the end-to-end flow:

**Screenshots and screen captures:**
Issue Rename:
<img width="910" height="539" alt="Screenshot From 2026-08-04 11-12-56" src="https://github.com/user-attachments/assets/17716208-f933-4619-ae05-8e426e9a6a91" />

Pull Request Rename:
<img width="823" height="718" alt="image" src="https://github.com/user-attachments/assets/da6603ea-2128-4fb6-8b6d-ce69effa2153" />

<details>
<summary>Rename Template:</summary>

<img width="892" height="154" alt="image" src="https://github.com/user-attachments/assets/e1b57e77-fc47-4bb8-ab30-c977ed03c068" />
</details>
<details>
<summary>PresetUrlOption:</summary>

<img width="601" height="715" alt="Screenshot From 2026-08-03 18-23-13" src="https://github.com/user-attachments/assets/e30dac9a-5165-4c40-9f78-3062052dfdc5" />
</details>

<details>
<summary>Doc Change:</summary>

<img width="786" height="147" alt="image" src="https://github.com/user-attachments/assets/64d0f2d4-73dd-4f1c-a594-c559cbe752cd" />
</details>

<details>
<summary>Self-review checklist</summary>

- [x] [Self-reviewed](https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code) the changes for clarity and maintainability
      (variable names, code reuse, readability, etc.).

Communicate decisions, questions, and potential concerns.

- [x] Explains differences from previous plans (e.g., issue description).
- [x] Highlights technical choices and bugs encountered.
- [x] Calls out remaining decisions and concerns.
- [x] Automated tests verify logic where appropriate.

Individual commits are ready for review (see [commit discipline](https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html)).

- [x] Each commit is a coherent idea.
- [x] Commit message(s) explain reasoning and motivation for changes.

Completed manual review and testing of the following:

- [ ] Visual appearance of the changes.
- [ ] Responsiveness and internationalization.
- [ ] Strings and tooltips.
- [x] End-to-end functionality of buttons, interactions and flows.
- [x] Corner cases, error conditions, and easily imagined bugs.
</details>

Niloth-p commented:
Please go through the [guide on submitting a pull request](https://zulip.readthedocs.io/en/latest/contributing/reviewable-prs.html) to learn how to present your proposed changes to Zulip, so that your work can be reviewed.

DhruvShetty22 commented:
@Niloth-p  sorry for the inconvenience, I've squashed the commit history. Please let me know if some other changes are required .

DhruvShetty22 committed:
webhooks: Add helper function to rename topics with history.

Fixes #34074

DhruvShetty22 committed:
webhooks/github: Rename topic when Pull Request title is edited.

DhruvShetty22 commented:
> @DhruvShetty22 I tested this locally. For pull request edit events, the topic is renamed correctly and the existing history is moved as expected. However, for issue edit events, a new topic is created instead of moving the existing conversation to the renamed topic. <img alt="image" width="1144" height="393" src="https://private-user-images.githubusercontent.com/190011583/531725205-4a03680d-71a4-4c3e-b6d5-1d4e67f3676e.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3Njc1Mzg2MDQsIm5iZiI6MTc2NzUzODMwNCwicGF0aCI6Ii8xOTAwMTE1ODMvNTMxNzI1MjA1LTRhMDM2ODBkLTcxYTQtNGMzZS1iNmQ1LTFkNGU2N2YzNjc2ZS5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjYwMTA0JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI2MDEwNFQxNDUxNDRaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT00OWZlODYwMzEwZGYzZTZiZGUxZWMyMmZmNWE5YmMwZDAzZmNiZjRlNWYyNWM1NDk1MTU1YWI3Y2RiNmZmMTViJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.OsA_XNrI3LDLx0B8yGKSkx_OOF2KPEAfZm1L1t-GwnE">

Hey it seems to work here

<img width="1092" height="651" alt="image" src="https://github.com/user-attachments/assets/1cd929fb-98f5-4656-8eef-40bd962c92cd" />

Probably you did not select issues as an event while adding the webhook from github.

Pritesh-30 commented:
@DhruvShetty22 I have tested it , currently its working as we expect. I still don't know why it wasn't working yesterday, the new branch and previous has no diff still. A small change then you can tag `Niloth-p` for review.

Pritesh-30 commented on zerver/webhooks/github/view.py:
```diff
@@ -1165,6 +1202,14 @@ def api_github_webhook(
+    if header_event == "issues" and payload.get("action", "").tame(check_string) == "edited":
+        stream_name = request.GET.get("stream")
+        if not stream_name and user_profile.default_sending_stream:
+            stream_name = user_profile.default_sending_stream.name  # nocoverage
```
Since we need it later for discussions also ,I thought it would be better to extract stream logic . I guess we should discuss it with @Niloth-p

Niloth-p commented on zerver/lib/webhooks/common.py:
```diff
@@ -342,3 +348,56 @@ def validate_webhook_signature(
+    stream_name: str,
+    old_topic: str,
+    new_topic: str,
+) -> None:
```
It would be nice if we could return bool - return True if the topic rename was successfully done.

Niloth-p commented on zerver/lib/webhooks/common.py:
```diff
@@ -342,3 +348,56 @@ def validate_webhook_signature(
+        target_message=last_message,
+        message_edit_request=edit_request,
+        send_notification_to_old_thread=False,
+        send_notification_to_new_thread=False,
```
I'm not sure if we shouldn't be sending a notification. The order would also matter if we did send it though.
We could also consider passing this in as an argument to the function.

But, it's fine for the moment. Just commenting for future reviewers.

laurynmm replied:
Probably worth having a discussion about this in the integrations CZO channel.

Niloth-p reviewed (commented):
Thanks for all the work, @DhruvShetty22! The logic for the common function looks good to me.
Thanks for the manual testing and reviews, @Pritesh-30!

Niloth-p commented on zerver/webhooks/github/view.py:
```diff
@@ -1131,7 +1222,7 @@ def api_github_webhook(
         request=request,
         payload=payload,
-        include_title=user_specified_topic is not None,
+        include_title=user_specified_topic is not None,  # nocoverage
```
Why is this edit required?

Niloth-p commented on zerver/tests/test_webhooks_common.py:
```diff
@@ -199,6 +200,37 @@ def test_check_send_webhook_message_returns_id(self) -> None:
+        with patch("zerver.lib.webhooks.common.do_update_message") as mock_update:
+            check_topic_rename(user_profile, stream_name, old_topic, new_topic)
+
+            mock_update.assert_called_once()
```
Can we also add a test case for multiple messages?

Niloth-p reviewed (commented):
Thanks for the prompt updates!

It may be okay to call the new functions from `api_github_webhook` instead of `get_opened_or_update_pull_request_body`, I don't have a strong enough opinion on it. But we should definitely refactor and tighten the calls.

Niloth-p reviewed (commented):
Thanks for the prompt updates!

`github/tests.py` needs to be updated quite a bit more.

We don't necessarily have to do the PRs and issues in the same commit, when we're going to be updating the fixtures too. Just implement the function for one, then tweak it to pass the next in the second commit. But, make sure the function is sensible in both commits. 

While choosing the 2 events to capture and update the fixtures for - it maybe simpler to go with the opened and labeled events. Feel free to pick anything suitable.

Thanks for the work!

DhruvShetty22 committed:
webhooks/github: Rename topic when Issue title is edited.

Testing this requires replaying two events for the same issue: the
topic that issues__opened creates is the topic that
issues__edited_title has to move away from.

The existing fixtures could not express that, because they were
captured by different people from different repositories. So replaying
them in sequence produced two unrelated topics and never exercised a
rename. Both are therefore a fresh capture of a single issue being
opened and then retitled, rather than a rename of the old files.

Niloth-p commented on zerver/tests/test_webhooks_common.py:
```diff
@@ -200,6 +201,64 @@ def test_check_send_webhook_message_returns_id(self) -> None:
+        )
+        id3 = self.send_stream_message(
+            self.user_profile, self.stream_name, topic_name=old_topic, content="Msg 3"
+        )
```
```suggestion
        msg_ids = [
            self.send_stream_message(self.user, self.stream, topic_name="Old", content=f"Msg {i}")
            for i in range(3)
        ]
```

Niloth-p commented on zerver/tests/test_webhooks_common.py:
```diff
@@ -200,6 +201,64 @@ def test_check_send_webhook_message_returns_id(self) -> None:
+
+        for msg_id in [id1, id2, id3]:
+            msg = Message.objects.get(id=msg_id)
+            self.assertEqual(msg.topic_name(), new_topic)
```
```suggestion
        messages = Message.objects.filter(id__in=msg_ids)
        self.assertTrue(all(message.topic_name() == "New" for message in messages))
```

Niloth-p commented on zerver/webhooks/github/view.py:
```diff
@@ -1156,9 +1162,9 @@ def api_github_webhook(
         return json_success(request)
 
-    # Handle title renaming for PRs with edited titles
+    # Handle title renaming for PRs and Issues with edited titles
```
"topic renaming" - see https://github.com/zulip/zulip/pull/37183#discussion_r2678512828

Niloth-p commented on zerver/webhooks/github/tests.py:
```diff
@@ -252,10 +253,80 @@ def test_pull_request_edit_ignore_non_title_changes(self) -> None:
+                realm=self.test_user.realm, subject=truncated_old_topic
+            ).exists(),
+            f"Message should exist at truncated topic: {truncated_old_topic}",
+        )
```
Why can't we use `check_webhook`?

Niloth-p commented on zerver/tests/test_webhooks_common.py:
```diff
@@ -200,6 +201,64 @@ def test_check_send_webhook_message_returns_id(self) -> None:
+            check_topic_rename(self.user_profile, self.stream_name, "Ghost Topic", "New")
+            mock_update.assert_not_called()
+
+    def test_topic_rename_propagation(self) -> None:
```
NIT: A test name suggestion was also mentioned in https://github.com/zulip/zulip/pull/37183#discussion_r2678508845.

Niloth-p commented on zerver/webhooks/github/tests.py:
```diff
@@ -172,6 +174,84 @@ def test_issue_msg(self) -> None:
+        full_old_topic = TOPIC_WITH_PR_OR_ISSUE_INFO_TEMPLATE.format(
+            repo=repo_name, type="PR", id=pr_number, title=long_title
+        )
+        truncated_old_topic = truncate_topic(full_old_topic)
```
Take a look at the other tests in the file.
They don't construct the message or topic, just hardcode it.

Niloth-p reviewed (commented):
Thanks for the updates, @DhruvShetty22!
This does look better, but the tests still don't look optimized to me. And some of my previous comments seem to have been missed.

I understand that the PR has grown in scope - from implementing the helper function to using it in the GitHub integration to editing the existing fixtures. Thanks for sticking with it!

The commit messages now speak of a "title renaming" - that doesn't make sense. "Topic" is a keyword in Zulip, we're renaming the "topic".

DhruvShetty22 committed:
webhooks/github: Update fixture filenames to match event actions.

Niloth-p commented on zerver/webhooks/github/tests.py:
```diff
@@ -183,14 +181,79 @@ def test_issue_comment_pull_request_comment_msg(self) -> None:
+        self.assertFalse(
+            Message.objects.filter(realm=self.test_user.realm, subject=expected_topic).exists(),
+            "Old truncated topic should have been renamed.",
+        )
```
```suggestion
        old_topic = "public-repo / PR #1 This is a very long Pull request titl..."
        new_topic = "public-repo / PR #1 New Short Title"
        opened_message = "baxterthehacker opened [PR #1](https://github.com/baxterthehacker/public-repo/pull/1) from `baxterthehacker:changes` to `baxterthehacker:master`:\n\n~~~ quote\nThis is a pretty simple change that we need to pull into master.\n~~~"
        edited_message = (
            "baxterthehacker edited [PR #1](https://github.com/baxterthehacker/public-repo/pull/1)."
        )

        self.check_webhook("pull_request__opened", old_topic, opened_message)
        self.check_webhook("pull_request__edited_title", new_topic, edited_message)
        self.assertFalse(
            Message.objects.filter(realm=self.test_user.realm, subject=old_topic).exists()
        )
```

Niloth-p commented on zerver/webhooks/github/tests.py:
```diff
@@ -183,14 +181,79 @@ def test_issue_comment_pull_request_comment_msg(self) -> None:
+            "Old truncated topic should have been renamed.",
+        )
+
+    def test_pull_request_body_edited_not_renamed(self) -> None:
```
This test doesn't make sense to me.

Niloth-p commented on zerver/webhooks/github/tests.py:
```diff
@@ -183,14 +181,79 @@ def test_issue_comment_pull_request_comment_msg(self) -> None:
+        self.assertFalse(
+            Message.objects.filter(realm=self.test_user.realm, subject=expected_topic).exists(),
+            f"Old topic '{expected_topic}' should have been renamed.",
+        )
```
```suggestion
        old_topic = "zulip-test / issue #33 This is a very long Issue title to..."
        opened_message = "DhruvShetty22 opened [issue #33](https://github.com/DhruvShetty22/zulip-test/issues/33):\n\n~~~ quote\nIt looks like you accidentally spelled 'commit' with two 't's.\n~~~"
        new_topic = "zulip-test / issue #33 New Short Title."
        edited_message = "DhruvShetty22 edited [issue #33](https://github.com/DhruvShetty22/zulip-test/issues/33):\n\n~~~ quote\nIt looks like you accidentally spelled 'commit' with two 't's.\n~~~"

        self.check_webhook("issues__opened", old_topic, opened_message)
        self.check_webhook("issues__edited_title", new_topic, edited_message)
        self.assertFalse(
            Message.objects.filter(realm=self.test_user.realm, subject=old_topic).exists()
        )
```

Niloth-p reviewed (commented):
Apologies for the delay, I appreciate the patience.
Thanks for the updates! The tests look much better now.
There may or may not be more optimal ways to update those fixtures, but this is acceptable for me. Thanks for arriving at this approach without waiting for my response.

Some of the changes in the last commit need to be moved to the third commit - the ones from my previous review - function rename, comment mentioning "title" instead of topic.

Please update the commit messages like this:
- Drop the commit message bodies of the last 2 commits.
- Move the Fixes line to the first commit itself.
- Use `webhooks/github` instead of `github`for the last 3 commits.

The rest LGTM!

DhruvShetty22 commented:
@Niloth-p  Done! Updated the commit history and changed the tests as per the suggestion. Honestly learned so much 
(especially patience). Thank you for the guidance :)

DhruvShetty22 commented:
@timabbott this has been in integration review for quite sometime now, can you please take a look at this.

alya commented:
@DhruvShetty22 No need to bump once the PR is the integration review stage; @timabbott will review when it reaches the top of his priorities list.

DhruvShetty22 commented:
Resolved Conflict and also updated `is_nontrivial_move=True,`  in  `StreamMessageEditRequest` in the helper function(which was added in bbf9b0eb17) as this is not a case only change ,
The test.py changes were added in patches to maintain proper commit history.

DhruvShetty22 commented:
@Niloth-p @laurynmm Can we have someone do the integration review on this one please?

laurynmm commented:
Sure. Let me put this one on my integration review list.

laurynmm commented on zerver/lib/webhooks/common.py:
```diff
@@ -419,3 +423,59 @@ def get_service_api_data(
+        orig_stream=stream,
+    )
+
+    do_update_message(
```
This is not checking if the user has the ability to do this message edit request. I'm pretty sure that we want to be calling `check_update_message`, which calls this function.

laurynmm commented on zerver/webhooks/github/view.py:
```diff
@@ -1237,6 +1275,15 @@ def api_github_webhook(
+        header_event == "pull_request"
+        and payload.get("action", "").tame(check_string) == "edited"
+        and "title" in payload.get("changes", {})
+        and stream
```
I feel like having the stream check earlier here would be better, because if it's not a channel message, then there's no topic to rename.

laurynmm commented on zerver/webhooks/github/view.py:
```diff
@@ -1237,6 +1275,15 @@ def api_github_webhook(
+        handle_topic_rename(user_profile, header_event, payload, stream)
+
     event = get_zulip_event_name(header_event, payload, branches)
     if event is None:
```
Hmm... why are we moving messages before checking if there's an event that we're handling here?

laurynmm reviewed (commented):
@DhruvShetty22 - Thanks for working on this issue!

I think there are a couple of major things that haven't quite been considered in these changes:
1. There are no permission checks happening in the `check_topic_rename` helper that has been created.
2. In the GitHub implementation, the topic move is happening before we've checked whether this is an event that we're going to handle/send a message for at all.

Also, looking for conversations about this issue in CZO, it looks like there was a mention about potentially having a way to disable this feature, which never really got discussed: [#integrations > git: auto resolve issue/pr topic on specific events](https://chat.zulip.org/#narrow/channel/127-integrations/topic/git.3A.20auto.20resolve.20issue.2Fpr.20topic.20on.20specific.20events/with/2132351).

So, I think there needs to be more testing and consideration around corner cases (for example, what happens if a Zulip user resolves or edits the topic manually?) before this is ready for integration review, as well as further community discussion around breadcrumb messages and enabling/disabling the feature.

I think it would be good to have your mentors and @Niloth-p do another round of reviews once you've worked on the above issues.

DhruvShetty22 commented:
Thank you for the Review!
I am not sure if we also want to disable this feature while it make sense to disable the auto resolve, when do we want to not retain the topic history? 

Thank you for spotting the missing permission checks @laurynmm switched to using check_update_message() which i has the permission check we need.

Reordered the stream check and making sure a message would actually be sent beforehand

* Added a new Commit to Enable/Disable this (only for github) with a discussion started here [#integrations > Support renaming topics by webhook integrations @ 💬](https://chat.zulip.org/#narrow/channel/127-integrations/topic/Support.20renaming.20topics.20by.20webhook.20integrations/near/2476338)

@Aditya8840 Tagging you for review

DhruvShetty22 commented:
@Niloth-p Tagging you for a review.

DhruvShetty22 commented:
@zulipbot add "maintainer review"

DhruvShetty22 commented:
@zulipbot remove "mentor review"

laurynmm commented:
@PieterCK would you be up for doing maintainer review here?

DhruvShetty22 commented:
@PieterCK Gentle Ping.

PieterCK commented on zerver/webhooks/github/view.py:
```diff
@@ -1232,6 +1232,7 @@ def api_github_webhook(
     ignore_private_repositories: Json[bool] = False,
     include_repository_name: Json[bool] = False,
     include_emoji_indicators: Json[bool] = True,
+    enable_topic_rename: Json[bool] = True,
```
Should the default actually be `False` to preserve existing behavior? Users with existing integrations who don't want this new behavior would need to re-set up their integration, which could be annoying. Worth discussing in CZO.

DhruvShetty22 replied:
I think we can leave the existing integration as it is by switching it to false and make only the new integration with the checkbox_enabled have this feature

PieterCK commented on zerver/webhooks/github/tests.py:
```diff
@@ -201,6 +201,29 @@ def test_issue_msg(self) -> None:
+        self.check_webhook("pull_request__edited_title", new_topic, edited_message)
+        self.assertFalse(
+            Message.objects.filter(realm=self.test_user.realm, subject=old_topic).exists()
+        )
```
I think we could better demonstrate the new behavior is correct by doing something like:
```diff
diff --git a/zerver/webhooks/github/tests.py b/zerver/webhooks/github/tests.py
index 098f8cc9bd..7da01a4529 100644
--- a/zerver/webhooks/github/tests.py
+++ b/zerver/webhooks/github/tests.py
@@ -210,10 +210,10 @@ class GitHubWebhookTest(WebhookTestCase):
             "baxterthehacker edited [PR #1](https://github.com/baxterthehacker/public-repo/pull/1)."
         )
         self.check_webhook("pull_request__opened", old_topic, opened_message)
+        self.assert_length(Message.objects.filter(realm=self.test_user.realm, subject=old_topic), 1)
         self.check_webhook("pull_request__edited_title", new_topic, edited_message)
-        self.assertFalse(
-            Message.objects.filter(realm=self.test_user.realm, subject=old_topic).exists()
-        )
+        self.assert_length(Message.objects.filter(realm=self.test_user.realm, subject=old_topic), 0)
+        self.assert_length(Message.objects.filter(realm=self.test_user.realm, subject=new_topic), 2)

     def test_issue_title_edit_moves_topic(self) -> None:
         old_topic = "zulip-test / issue #33 This is a very long Issue title to..."
```
We should also update similar tests.

PieterCK commented on zerver/webhooks/github/fixtures/issues__opened.json:
```diff
@@ -1,154 +1,197 @@
+        "type": "User",
+        "user_view_type": "public",
+        "site_admin": false
+    }
```
It is not obvious to me why these two fixtures are replaced. I think if we have to update existing fixtures instead of adding a new set of fixtures here, a little comment explaining why in the commit description would be helpful.

DhruvShetty22 replied:
It was added because old fixtures were named ambiguously:
Apart from editing tests it did the following thus i think the commit title explains it

```
(zulip-server) vagrant@vagrant:/home/shetty/Desktop/zulip/zulip$ git show 7c9f817a42cc2c83f823aa0b98ef018d5bdddee0
commit 7c9f817a42cc2c83f823aa0b98ef018d5bdddee0
Author: Dhruv Shetty <dhruvshetty2203@gmail.com>
Date:   Thu Jan 22 20:13:44 2026 +0000

    webhooks/github: Update fixture filenames to match event actions.

diff --git a/zerver/webhooks/github/fixtures/issues.json b/zerver/webhooks/github/fixtures/issues__opened.json
similarity index 100%
rename from zerver/webhooks/github/fixtures/issues.json
rename to zerver/webhooks/github/fixtures/issues__opened.json
diff --git a/zerver/webhooks/github/fixtures/pull_request__edited_with_body_change.json b/zerver/webhooks/github/fixtures/pull_request__edited_body.json
similarity index 100%
rename from zerver/webhooks/github/fixtures/pull_request__edited_with_body_change.json
rename to zerver/webhooks/github/fixtures/pull_request__edited_body.json
diff --git a/zerver/webhooks/github/fixtures/pull_request__edited.json b/zerver/webhooks/github/fixtures/pull_re
quest__edited_title.json
similarity index 100%
rename from zerver/webhooks/github/fixtures/pull_request__edited.json
rename to zerver/webhooks/github/fixtures/pull_request__edited_title.json
diff --git a/zerver/webhooks/github/tests.py b/zerver/webhooks/github/tests.py
index 90c563f0e2d..407ed3ff313 100644
--- a/zerver/webhooks/github/tests.py
+++ b/zerver/webhooks/github/tests.py

```
would be great if  you could  tell me what is not clear so that i can mention exactly that

PieterCK replied:
Sorry if my comment wasn't clear. To be specific, I was talking about the changes in commit `webhooks/github: Rename topic when Issue title is edited`. I think the reason fixture `issue__opened.json` was **replaced** with a new payload (not **renamed**) isn't obvious. So to make it easier for people later on to understand why it was changed, we should write the reason in that commit's description.

DhruvShetty22 replied:
Ahh Thankyou for the clarification
Adding the following note:-

> webhooks/github: Rename topic when Issue title is edited.

Testing this requires replaying two events for the same issue: the
topic that issues__opened creates is the topic that
issues__edited_title has to move away from.

The existing fixtures could not express that, because they were
captured by different people from different repositories. So replaying
them in sequence produced two unrelated topics and never exercised a
rename. Both are therefore a fresh capture of a single issue being
opened and then retitled, rather than a rename of the old files.

PieterCK reviewed (commented):
I posted a few comments here and in [#integrations > Support renaming topics by webhook integrations](https://chat.zulip.org/#narrow/channel/127-integrations/topic/Support.20renaming.20topics.20by.20webhook.20integrations/with/2502687)!

DhruvShetty22 committed:
webhooks/github: Report the old and new title on a title edit.

Editing a pull request or issue title sent a bare "edited"
notification, so with the topic now renamed underneath it, nothing in
the conversation explained why it moved. Report the old and new title
instead, which also drops the issue description that was previously
requoted on every title edit.

DhruvShetty22 committed:
webhooks/github: Make topic renaming on title edits opt-in.

Renaming the topic when a PR or issue title is edited would be a
surprising change for integrations that are already configured, whose
URLs carry no such parameter. Gate it behind an `enable_topic_rename`
URL option that defaults to off.

The option is surfaced as a checkbox that the integration URL modal
ticks by default and writes into the URL explicitly, so newly created
integrations still get the rename without any extra work.

DhruvShetty22 commented:
@PieterCK Thank you for the review!

I did a lot of changes to this which i found after testing once the commit adding the Rename template was added. 
I also found a time based restriction `check_time_limit_for_change_all_propagate_mode` which i've posted a message on CZO.

1) Added commit **webhooks/github: Report the old and new title on a title edit.** which provides the Rename Template.
2) Added ENABLE_TOPIC_RENAME as a PresetUrlOption attached SS and also doc changes
3) The stream based topic lookup was probably not the correct approach Thus for the `check_topic_rename` helper function i added moved from using  `access_stream_by_name` to `access_message` and using **sent_message_id** instead of **stream**

4) Improved Old tests based on these changes and added some new tests

PieterCK commented on zerver/lib/webhooks/common.py:
```diff
@@ -419,3 +429,44 @@ def get_service_api_data(
+    old_topic: str,
+) -> bool:
+    try:
+        sent_message = access_message(user_profile, sent_message_id, is_modifying_message=False)
```
So we mainly fetch this to look up the issue/PR notifications' `recipient_id`. Instead of calling `access_message`, I think we can fetch that by parsing the `stream` query from the integration URL and just do something like:
```python
    try:
        if stream_name_or_id.isdecimal():
            stream = get_stream_by_id_in_realm(stream_name_or_id, user_profile.realm.id)
        else:
            stream = get_realm_stream(stream_name_or_id, user_profile.realm.id)
    except StreamDoesNotExistError:
        return False
```

And then `stream_recipient_id=stream.recipient.id`.

PieterCK commented on docs/webhooks/incoming-webhooks-reference.md:
```diff
@@ -175,6 +175,16 @@ The currently configured preset URL options are:
+  and adds UI for the user to choose whether editing the title of a pull
+  request or issue also renames the topic that its notifications are sent
+  to, moving the messages already in that topic along with it. This option
+  is selected by default, and the `enable_topic_rename` boolean parameter
```
Shouldn't this say that this option is **disabled** by default to preserve the existing behavior?

PieterCK replied:
I think in the "Generate an integration URL" modal it doesn't matter what the default is for this field. But we should say that in the backend, it defaults to `False`.

PieterCK commented on zerver/lib/webhooks/common.py:
```diff
@@ -102,6 +106,12 @@ def build_preset_config(cls, config: PresetUrlOption) -> "WebhookUrlOption":
                     label="Exclude notifications from private repositories",
                     input_type="checkbox",
                 )
+            case PresetUrlOption.ENABLE_TOPIC_RENAME:
```
I think in the UI it'd be great if this field is only enabled when a channel is selected, similar to "Send all notifications to a single topic". And speaking of which, it should be disabled too if that that field is checked.

PieterCK commented on zerver/webhooks/github/tests.py:
```diff
@@ -201,6 +201,33 @@ def test_issue_msg(self) -> None:
+        )
+        self.assert_length(
+            Message.objects.filter(
+                realm=self.test_user.realm, subject="public-repo / PR #1 New Short Title"
```
I think we should at least use `TOPIC_WITH_PR_OR_ISSUE_INFO_TEMPLATE` to compute the topic for this.

DhruvShetty22 commented:
@PieterCK 


- Done, check_topic_rename accepts stream now . Also rewrote the PresetUrl description. 

I think we can get rid of handle_topic_rename function completely by passing old title to get_topic_based_on_type and directly calling 

```
    if (
        sent_message_id is not None
        and enable_topic_rename
        and user_specified_topic is None
        and header_event in ("pull_request", "issues")
        and payload["action"].tame(check_string) == "edited"
        and "title" in payload.get("changes", {})
    ):
        check_topic_rename(
            request,
            user_profile,
            get_topic_based_on_type(
                payload, event, payload["changes"]["title"]["from"].tame(check_string)
            ),
            topic_name,
        )
```

> I did a lot of changes to this which i found after testing once the commit adding the Rename template was added. I also found a time based restriction check_time_limit_for_change_all_propagate_mode which i've posted a message on CZO.


and i guess this got missed

PieterCK commented on zerver/lib/webhooks/common.py:
```diff
@@ -419,3 +423,55 @@ def get_service_api_data(
+
+@typed_endpoint
+def check_topic_rename(
+    request: HttpRequest,
```
This parameter might be a dead code.

DhruvShetty22 replied:
it was there because of the @typed_endpoints, i think it is better to pass stream directly through api_github_webhook

PieterCK commented on zerver/webhooks/github/view.py:
```diff
@@ -1042,13 +1043,13 @@ def is_pull_request_comment_event(payload: WildValue) -> bool:
 
 
-def get_topic_based_on_type(payload: WildValue, event: str) -> str:
+def get_topic_based_on_type(payload: WildValue, event: str, title: str | None = None) -> str:
```
I think this is confusing. We use this to compute a topic name based off the `payload` and `event`. But if we also pass in the ambiguous `title` parameter, it would somehow use it to compute the topic name?

I think we'd better off with a separate new helper function called something like `get_old_topic_from_rename_event`.

PieterCK commented on docs/webhooks/incoming-webhooks-reference.md:
```diff
@@ -175,6 +175,20 @@ The currently configured preset URL options are:
+  it. The option is only offered when the notifications go to a channel and
+  are not all sent to a single topic. In the webhook, the `enable_topic_rename`
+  boolean parameter defaults to `False`, so that integration URLs generated
+  before this option existed keep their original behavior.
```
We should mention the caveats of using `check_topic_rename` here too.

```

This setting uses `check_topic_rename` to move messages, which comes with these limitations ....
```
[#integrations > Support renaming topics by webhook integrations @ 💬](https://chat.zulip.org/#narrow/channel/127-integrations/topic/Support.20renaming.20topics.20by.20webhook.20integrations/near/2508971)

and

DhruvShetty22 replied:
>   Renaming uses `check_topic_rename`, which respects the same permissions
  and time limits as a user moving a topic. It skips the move if the bot
  cannot [move messages](https://zulip.com/help/restrict-moving-messages)
  between topics, or if the conversation is older than the organization's
  time limit for moving messages, 7 days by default. Skipping is silent, so
  the new notification starts a fresh topic while the earlier messages stay
  under the old title.

PieterCK commented on zerver/webhooks/github/view.py:
```diff
@@ -1057,7 +1057,7 @@ def get_topic_based_on_type(payload: WildValue, event: str, title: str | None =
             type=type_for_topic,
             id=payload["issue"]["number"].tame(check_int),
-            title=payload["issue"]["title"].tame(check_string),
+            title=payload["issue"]["title"].tame(check_string) if title is None else title,
```
We should use the `get_old_topic_from_rename_event` function for this too.

DhruvShetty22 commented:
@PieterCK 
Thank you for the quick review!!! (I hope i get this merged before gsoc endterm)
I would definitely be up for the follow up PRs this requires for completion like the time-restriction etc...

Changes:
* Restructured commit as per your  comments (Looks great now :)
* Added the caveats to the documentation for enable_topic_rename
* Added helper get_old_topic_from_rename_event
* Added a test for time-restriction.

PieterCK commented on zerver/lib/webhooks/common.py:
```diff
@@ -419,3 +430,57 @@ def get_service_api_data(
+) -> bool:
+    if stream is None:
+        return False
+
```
I think `stream` can be a plain parameter instead. If we want to make sure callers don't mix up the parameters, we should also include the `old_topic_name` and `new_topic_name` as part of the mandatory keyword arguments too.

We also want it to always be `str`, and do the `if stream:` check before we call `check_topic_rename`. And I think we should start calling it a `channel` here.

```suggestion
def check_topic_rename(
    user_profile: UserProfile,
    *
    old_topic_name: str,
    new_topic_name: str,
    channel: str,
) -> bool:
```

PieterCK commented on zerver/lib/webhooks/common.py:
```diff
@@ -419,3 +430,57 @@ def get_service_api_data(
+        # different channel, a subsequent topic rename may still trigger this
+        # check, but there will be no messages in the old topic.
+        # In that case, we don't want to rename the topic.
+        return False
```
```suggestion
    # This could happen if the user reconfigures their integration to send
    # notifications to a different channel and, right after that, triggers
    # an event that renames the topic. Currently we have no way of handling
    # that case, and PresetUrlOption.ENABLE_TOPIC_RENAME doesn't claim to be
    # able to move notifications across channels.
    if last_message_in_old_topic is None:
        return False
```

PieterCK commented on web/src/integration_url_modal.ts:
```diff
@@ -210,6 +211,15 @@ export function show_generate_integration_url_modal(api_key: string): void {
             }
         }
 
+        function update_topic_rename_ui(): void {
```
This logic should be part of this loop:

```javascript
            if (url_options) {
                for (const option of url_options) {
                    const {key, input_type} = option;
```
In the `update_url` function together with the custom logic for the all the other `PresetUrlOption`.

PieterCK commented on web/src/integration_url_modal.ts:
```diff
@@ -350,8 +362,11 @@ export function show_generate_integration_url_modal(api_key: string): void {
+                        if ($checkbox.is(":disabled")) {
+                            continue;
+                        }
+                        if ($checkbox.is(":checked")) {
```
Hmm can you explain why we need this change?

DhruvShetty22 replied:
That was there because the option was declared checkbox_enabled earlier, without the skip it would have written enable_topic_rename=true into a URL.
I switched to unchecked box now

```
if (key === PresetUrlOption.ENABLE_TOPIC_RENAME) {
    const disable_topic_rename =
        stream_id === direct_messages_option.unique_id ||
        $override_topic.is(":checked");
    set_input_disabled_state($checkbox, disable_topic_rename);
    if (disable_topic_rename) {
        $checkbox.prop("checked", false);
    }
}
const is_checked = $checkbox.is(":checked");

```

DhruvShetty22 commented:
@PieterCK 
Fixed the bugs on Integration URL model and using **channel**  as a argument now

PieterCK commented on docs/webhooks/incoming-webhooks-reference.md:
```diff
@@ -175,6 +175,28 @@ The currently configured preset URL options are:
+  between topics, or if the conversation is older than the organization's
+  time limit for moving messages, 7 days by default. Skipping is silent, so
+  the new notification starts a fresh topic while the earlier messages stay
+  under the old title.
```
I think the commit description for third commit could be a lot simpler. This documentation already explains what the feature does and other details. No need to repeat them again in the commit description.

PieterCK commented on zerver/webhooks/github/view.py:
```diff
@@ -125,6 +127,8 @@ def get_opened_or_update_pull_request_body(helper: Helper) -> str:
     description = None
     changes = payload.get("changes", {})
+    if "title" in changes:
+        return get_title_edited_body(helper, pull_request, "PR")
```
So in the original `issues__edited_title.json` it shows that it's possible for a `title` and `body` change event to happen in the same payload and produce `"changes"` field like this:
```json
  "changes": {
    "body": {
      "from": ""
    },
    "title": {
      "from": "New Issue edited"
    }
  },
```
So if that fixture is legit, this would skip the body change detail. Can you confirm whether that kind of payload is possible?

DhruvShetty22 replied:
Hmm..... Github sends 
  ```
"changes": {
    "body": {
      "from": ""
    },
```
for every title edit but an issue discription only change has only body in changes 
I dont think it is possible to edit both simultaneously  from the UI.

PieterCK commented on zerver/lib/webhooks/common.py:
```diff
@@ -419,3 +430,55 @@ def get_service_api_data(
+            topic_name=old_topic_name,
+        )
+        .order_by("-date_sent")
+        .first()
```
`get_latest_message_for_user_in_topic` does this to get the last message in a topic:

```python
      messages_for_topic(realm_id, recipient_id, topic_name)
      .values_list("id", flat=True)
      .last()
```

Should we copy that here?

PieterCK commented on zerver/lib/webhooks/common.py:
```diff
@@ -102,6 +107,12 @@ def build_preset_config(cls, config: PresetUrlOption) -> "WebhookUrlOption":
+            case PresetUrlOption.ENABLE_TOPIC_RENAME:
+                return cls(
+                    name=config.value,
+                    label="Rename topics when a pull request or issue title is edited",
```
We should rephrase this now that we frame `PresetUrlOption.ENABLE_TOPIC_RENAME` as a generic config.

Maybe something like:
```suggestion
                    label="Rename topics when a title is edited",
```

Not sure.

DhruvShetty22 replied:
I think this  should work label="Rename topics when a title is edited",

other suggestion could be 
"Rename topics and move their messages when a title changes"
"Keep topic names in sync with the third-party service"

PieterCK commented:
@DhruvShetty22 -- Thanks for making those changes. I posted a few more comments!

DhruvShetty22 commented:
@PieterCK 

With respect to this

>       messages_for_topic(realm_id, recipient_id, topic_name)
>       .values_list("id", flat=True)
>       .last()


I think it is better to use 
```
first_message_id_in_old_topic = (
        messages_for_topic(
            realm_id=user_profile.realm_id,
            stream_recipient_id=stream.recipient_id,
            topic_name=old_topic_name,
        )
        .values_list("id", flat=True)
        .first()
    )
    

```    
instead of last message I used .first(). That helper(`get_latest_message_for_user_in_topic`) wants the newest message for an access check, but here the newest message in the old topic can be the notification we just sent for this very event. messages_for_topic matches subject__iexact, so on a rename that **only changes the case of the title**, the notification we posted to the new topic is also in the old one. Anchoring on it makes check_update_message drop the topic edit , so we dont see the  MOVED marker. 
propagate_mode="change_all" moves the whole topic regardless of the anchor.

PieterCK commented:
@DhruvShetty22 -- Thanks for making those changes!

> So, I think there needs to be more testing and consideration around corner cases (for example, what happens if a Zulip user resolves or edits the topic manually?) before this is ready for integration review, as well as further community discussion around breadcrumb messages and enabling/disabling the feature.

Ok, I think we haven't thought about what to do if the topic is resolved/edited by the user. But I'm also not sure if that's something we absolutely need for the first version of this feature. I'm leaning towards putting that off to a follow up issue since this PR's size and discussion history are getting unwieldy.

On breadcrumb messages when renaming topics, we discussed this in [#integrations > Support renaming topics by webhook integrations](https://chat.zulip.org/#narrow/channel/127-integrations/topic/Support.20renaming.20topics.20by.20webhook.20integrations/with/2132305). So far, the decision is to update the "title changed" events to be clearer that an issue/PR's title has been changed and no notification bot message on top of that. This is still kinda tentative. If we think otherwise, the changes should be simple since we only need to toggle a parameter.

I haven't manually tested these changes with a live GitHub integration but the screenshots look good to me. I've manually checked the "Generate integration URL" modal for potential regressions and to test the new URL option.

@laurynmm -- Overall, I think these changes looks good to me. Would you be up for doing integration review here? Or I can continue the maintainer review if you think the items I mentioned above should be addressed in this PR.

laurynmm commented:
Started [#backend > allowing webhooks to rename topics - limit size of move?](https://chat.zulip.org/#narrow/channel/3-backend/topic/allowing.20webhooks.20to.20rename.20topics.20-.20limit.20size.20of.20move.3F/with/2515836) to just check in about whether we want to add some sort of limit to the number of messages moved in the helper function.
